File: ts/web3js/src/instructions/eata/undelegateEphemeralAtaPermission.ts
import {
  PublicKey,
  TransactionInstruction,
  AccountMeta,
} from "@solana/web3.js";
import {
  EATA_PROGRAM_ID,
  MAGIC_PROGRAM_ID,
  MAGIC_CONTEXT_ID,
  PERMISSION_PROGRAM_ID,
} from "../../constants";
import {
  ephemeralAtaPdaWithBumpFromOwnerAndMint,
  permissionPdaFromAccount,
} from "../../pda";

export interface UndelegateEphemeralAtaPermissionAccounts {
  payer: PublicKey;
  user: PublicKey;
  mint: PublicKey;
}

export function createUndelegateEphemeralAtaPermissionInstruction(
  accounts: UndelegateEphemeralAtaPermissionAccounts,
): TransactionInstruction {
  const [ephemeralAta] = ephemeralAtaPdaWithBumpFromOwnerAndMint(
    accounts.user,
    accounts.mint,
  );
  const permission = permissionPdaFromAccount(ephemeralAta);

  const keys: AccountMeta[] = [
    { pubkey: accounts.payer, isWritable: false, isSigner: true },
    { pubkey: ephemeralAta, isWritable: true, isSigner: false },
    { pubkey: permission, isWritable: true, isSigner: false },
    { pubkey: PERMISSION_PROGRAM_ID, isWritable: false, isSigner: false },
    { pubkey: MAGIC_PROGRAM_ID, isWritable: false, isSigner: false },
    { pubkey: MAGIC_CONTEXT_ID, isWritable: true, isSigner: false },
  ];

  const buffer = Buffer.alloc(1);
  buffer[0] = 8; // discriminator

  return new TransactionInstruction({
    programId: EATA_PROGRAM_ID,
    keys,
    data: buffer,
  });
}

File: rust/sdk/src/spl/builders/undelegate_ephemeral_ata_permission.rs
use crate::{
    access_control::structs::Permission,
    compat,
    consts::{ESPL_TOKEN_PROGRAM_ID, MAGIC_CONTEXT_ID, MAGIC_PROGRAM_ID, PERMISSION_PROGRAM_ID},
    spl::{EphemeralAta, EphemeralSplDiscriminator},
};

pub struct UndelegateEphemeralAtaPermissionBuilder {
    pub payer: compat::Pubkey,
    pub user: compat::Pubkey,
    pub mint: compat::Pubkey,
}

impl UndelegateEphemeralAtaPermissionBuilder {
    #[inline(always)]
    pub fn instruction(&self) -> compat::Instruction {
        let (eata, _eata_bump) = EphemeralAta::find_pda(&self.user, &self.mint);
        let (permission, _permission_bump) = Permission::find_pda(&eata);

        compat::Instruction {
            program_id: ESPL_TOKEN_PROGRAM_ID,
            accounts: vec![
                compat::AccountMeta::new_readonly(self.payer, true),
                compat::AccountMeta::new(eata, false),
                compat::AccountMeta::new(permission, false),
                compat::AccountMeta::new_readonly(PERMISSION_PROGRAM_ID, false),
                compat::AccountMeta::new_readonly(MAGIC_PROGRAM_ID, false),
                compat::AccountMeta::new(MAGIC_CONTEXT_ID, false),
            ],
            data: vec![EphemeralSplDiscriminator::UndelegateEphemeralAtaPermission as u8],
        }
    }
}
